### PR TITLE
fix(@ember/object/core): add variadic `...args` to `init`

### PR DESCRIPTION
**Pull request by buschtoens**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:
  - https://guides.emberjs.com/release/applications/services/#toc_defining-services
  - https://discordapp.com/channels/480462759797063690/484421406659182603/524230021435359232
- ~~Increase the version number in the header if appropriate.~~
- ~~If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.~~

---

In the guides linked above it is recommended to _always_, even for services, pass along all arguments:

```js
export default Service.extend({
  init() {
    this._super(...arguments);
    // ...
  }
});
```

Which translates to:

```ts
export default class MyService extends Service {
  init(...args: any[]) {
    super.init(...args);
    // ...
  }
});
```

But this currently causes the following type mismatch:

image

This PR fixes the root cause by making `CoreObject#init` variadic.

### CODE REVIEW

**Review by mfeckie — APPROVED** [review]

**Review by mike-north — CHANGES_REQUESTED** [review]
While I understand that this makes the examples in the guides type-check, before this is merged I need more certainty around `init()` receiving args being part of Ember's public API.

> In the guides linked above it is recommended to always, even for services, pass along all arguments.

@buschtoens can you quote the relevant text directly from the page you linked to in the guides? I can't find the text you seem to be referring to. It's also worth mentioning that "passing all arguments to super" doesn't necessarily mean that it _must_ be done via a spread operator and the `arguments` object.

If [the public API of Ember describes `init()` as not receiving any args](https://emberjs.com/api/ember/3.6/classes/CoreObject/methods/init?anchor=init), it would seem that for _typescript users_, 

```ts
class {
	init() {
	   super.init();
	}
}
```

is the appropriate pattern to use for hooks that receive no arguments.

**Comment by dwickern** [comment]
@mike-north check the notes for the API docs you linked:
> NOTE: If you do override init for a framework class like Ember.View, be sure to call this._super(...arguments) in your init declaration! If you don't, Ember may not have an opportunity to do important setup work, and you'll see strange behavior in your application.

**Comment by mike-north** [comment]
>  @mike-north check the notes for the API docs you linked:
> 
> NOTE: If you do override init for a framework class like Ember.View, be sure to call this._super(...arguments) in your init declaration! If you don't, Ember may not have an opportunity to do important setup work, and you'll see strange behavior in your application.

Thanks for catching this! 

This may be a good example to kick off a discussion about docs for TS vs JS users. The "always use `...arguments`" rule of thumb is a safe bet for JS developers -- in this case it does not hold up well for TS consumers.

There are a bunch of zero-argument hooks, and if we decide to treat them differently, they should all be treated consistently. I think this warrants a discussion around correctness and these types, specifically figuring out whether we align with

* API docs
* Conventions mentioned in the guides (i.e, "always do `this._super(...arguments);`)
* Ember source code

In this situation at least two of these are conflicting _when type information used with `typescript@latest` is taken into account_, but are in alignment from a JS perspective.

**Comment by buschtoens** [comment]
> There are a bunch of zero-argument hooks, and if we decide to treat them differently, they should all be treated consistently.

:+1: on consistency.

I think two further questions to ask are:

- What up- or downsides are there to adding `...arguments` everywhere and in the types and code?
  - Forward stability, if some zero-argument hooks ever receive arguments.
  - Consistent pattern: Just always pass `...arguments`. Thanks to IntelliSense, we don't need a consistent pattern though.
    - Derived constructors / `init` that explicitly demand a certain signature or no arguments at all can still opt out, by overriding the constructor.
  - Code bloat for functionality that likely will never be used.
- How likely is it that zero-argument hooks (or super methods in general) will be updated to receive arguments in the future?
  - Same question might be asked for methods that already receive arguments: Will they ever receive more arguments in the future?
  - `init` / the constructor might be a bit of a special snowflake here, because of setup work that might be performed by a factory function now, that could be inlined into the super constructor.

The more I think about the arguments (HA!) brought forth by @mike-north, the more I'm leaning towards not merging this PR and instead making sure all parameters are typed 💯 correctly.

**Comment by buschtoens** [comment]
@mike-north has elaborated a bit further on this [on Discord](https://discordapp.com/channels/480462759797063690/484421406659182603/524495974584811532).

**Pull request closed without merging**
